addons/capetrader/capetrader.lua

```diff
@@ -0,0 +1,495 @@
+--[[Copyright © 2016, Lygre, Burntwaffle@Odin
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of CapeTrader nor the
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL Lygre, Burntwaffle@Odin BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.]]--
+
+_addon.name = 'CapeTrader'
+_addon.author = 'Lygre, Burntwaffle'
+_addon.version = '1.0.2'
+_addon.commands = {'capetrader', 'ct'}
+
+require('luau')
+require('pack')
+require('sets')
+require('functions')
+require('strings')
+local res = require('resources')
+local packets = require('packets')
+local augPaths = require('allAugPaths')
+local maxAugMap = require('maxAugMap')
+local extData = require('extdata')
+local jobToCapeMap = require('jobToCapeMap')
+
+local playerIndex = nil
+local gorpaID = 17797273
+local gorpaTargetIndex = 153
+local gorpaMenuID = 387
+local mhauraID = 249
+local currentCape
+local pathName = nil
+local pathIndex
+local pathItem = nil
+local inventoryBagNumber = 0
+local maxAmountDustAndThread = 20
+local maxAmountSapAndDye = 10
+local capeIndex
+local augmentItemIndex
+local capeHasBeenPrepared = false
+local currentlyAugmenting = false
+local firstTimeAug = false
+local timesAugmentedCount = 0
+local numberOfTimesToAugment = 0
+local dustSapThreadTradeDelay = 1
+local dyeTradeDelay = 2
+local tradeReady = false
+local threadIndex = 1
+local dustIndex = 2
+local dyeIndex = 3
+local sapIndex = 4
+local maxAugKey = nil
+local zoneHasLoaded = true
+local inventory = nil
```

Comment:
You rely on a lot of global state. This is a simple way to implement certain mechanisms, but is also prone to some issues. For example, you always have to remember to clean up the state wherever you leave certain functions. This makes those functions harder to design, since you can really only have one exit route.

Instead I'd consider passing more things as arguments. One simple example is the `getCapeIndex` function. It uses the `currentCape` variable, which is defined globally. Instead, you can simply pass the value to the function where it's called inside `checkCapeCount`. Which, in turn, can also receive it from its calling function. This both avoids the need to clean up, since it doesn't persist by default, as well as allows it to trace easily which functions need which values, since you see it in their argument list. That helps others (like me :D) to understand the code better and will help you if you decide to refactor it in the future.

The same can be applied to many other variables.

On the other hand, certain variables make sense to handle globally, namely those that are either inefficient to calculate (like player ID/index or zone info). Those can be stored globally and certain events can be registered which shall update those variables (like login/logout/zoning).